### ISSUE TITLE
Huge increase in rps after gem update

### ISSUE THREAD

**Issue by Mathiou04** [DataDog/dd-trace-rb#413]
Hi there,

after upgrading our gem from 0.7.1 to 0.11.4 on one of our services we've seen a huge increase in the rps metrics (rails controller, sidekiq and database), ranging between +50% to +100%.
Here is an example graph of what happened since the deployment:
<img width="417" alt="capture d ecran 2018-05-07 a 11 56 35" src="https://user-images.githubusercontent.com/14098361/39696527-47295b64-51ee-11e8-80bd-cd484e8aab9b.png">

I went through the change logs but didn't find a bugfix that could explain this (maybe #133 ?).
Any idea what is happening?

**Comment by palazzem** [comment]
Hello @Mathiou04! we did a lot of changes between 0.7.x and 0.11.x, probably increasing the visibility and so the size of the payload. These changes should not increase the number of request, so would be great if you can provide us more details about your stack. Can you share your stack and Datadog Tracer configuration?

Also is it happening only on one service or all your services are impacted?

**Comment by Mathiou04** [comment]
Hi @palazzem, thanks for the quick reply. I added a few more details above.
Concerning the stack, we are using rails 4.2.4 with redis/postgres/sidekiq 4.2.6.

Here is the current datadog config:
```
Datadog.configure do |c|
    # Tracer
    c.tracer hostname: '***', port: 8126, env: ENV['RACK']

    c.use :rails,
          service_name: 'web',
          controller_service: 'controller',
          cache_service: 'cache',
          database_service: 'postgres',
          distributed_tracing: true

    c.use :redis

    c.use :http, distributed_tracing: true

    c.use :sidekiq, service_name: 'worker'
  end
```

Here is what we had before:
```
Rails.configuration.datadog_trace = {
      auto_instrument: true,
      auto_instrument_redis: true,
      default_service: 'web',
      default_cache_service: 'cache',
      default_database_service: 'postgres',
      default_controller_service: 'controller',
      env: ENV['RACK'],
      trace_agent_hostname: '***',
      trace_agent_port: 8126
  }

  Datadog::Monkey.patch_module(:http)

  # Worker tracing
  require 'ddtrace/contrib/sidekiq/tracer'

  Sidekiq.configure_server do |config|
    config.server_middleware do |chain|
      chain.add(
          Datadog::Contrib::Sidekiq::Tracer,
          sidekiq_service: 'worker'
      )
    end
end
```

I didn't update any other service yet. I will do it today and I will let you know what happens.

**Comment by delner** [comment]
Is it possible because the number of traced requests went up, so did latency? Such that it is settling around a more accurate number?

**Comment by Mathiou04** [comment]
@delner That's what I thought first but when you think about it, I am not sure it makes sense: even if you have twice more request, you're still looking at the slowest 5% when you look at the 95 percentile. Or maybe it would mean that the gem was not discarding requests randomly, but the slower it was the more change it would get to be discarded.

@palazzem I did upgrade the other service that was using the outdated version of the gem but the change is less obvious (as this service gets a lot less requests...). I think the change is still there but I am gathering more data in order to have a clearer graph.

**Comment by delner** [comment]
I'm not 100% about this, but its possible that traces weren't being sampled in the same proportions as they are now.

What version of the agent are you running?

**Comment by Mathiou04** [comment]
@delner Here is an extract from version-manifest.json:
```
"build_version": "5.14.0",
"build_git_revision": "ecde6c70b36124c70b3452f0038dbdcab59a0ea1"
```

@palazzem Not seeing the same increase on the other service. Maybe worth mentionning that the service seeing the increase also has the gem dogstatsd-ruby installed?

It is very clear that the change was only in the metrics, not in the load. Here is a comparison between the datadog rps graph and the aws cpu utilization graph on the exact same period:
![capture d ecran 2018-05-09 a 14 44 59](https://user-images.githubusercontent.com/14098361/39815221-9855169a-5397-11e8-8776-55d55754260d.png)

![capture d ecran 2018-05-09 a 14 44 49](https://user-images.githubusercontent.com/14098361/39815219-96da718e-5397-11e8-9112-97631d8e1e87.png)

**Comment by Mathiou04** [comment]
@palazzem Update on the other service, not seeing the huge increase in rps, but still seeing a change in the latency (this time a decrease) :
<img width="396" alt="capture d ecran 2018-05-11 a 16 30 02" src="https://user-images.githubusercontent.com/14098361/39929625-ce7a06ac-5538-11e8-9ee6-b58fc059d62b.png">

**Comment by Mathiou04** [comment]
![capture d ecran 2018-05-23 a 18 31 14](https://user-images.githubusercontent.com/14098361/40438113-8033a254-5eb7-11e8-9ab9-6b0f7aa27f03.png)

Just noticed we had the same issue with redis monitoring. The new graph makes a lot more sense than the first, but I am still interested in an explanation.

It looks like some commands magically appeared after the update:
![capture d ecran 2018-05-23 a 18 35 02](https://user-images.githubusercontent.com/14098361/40438286-0528a054-5eb8-11e8-8078-e7fd4582e69e.png)
![capture d ecran 2018-05-23 a 18 35 38](https://user-images.githubusercontent.com/14098361/40438301-136e6464-5eb8-11e8-9532-ef3e04d9312b.png)

**Comment by delner** [comment]
@Mathiou04 I checked with our team about how this number might have been affected. I think the most likely explanation for this is that we're tracing a lot more things in 0.11.x vs 0.7.x, and the quantity of traces can greatly affect this RPS metric you're seeing.

Are you seeing any other strong indicators of performance problems? (Other than RPS metrics?) The CPU graph you linked above looks congruent with the RPS, if Im not misreading it.

**Comment by Mathiou04** [comment]
@delner Thanks for your feedback.
Actually, there is no performance degradation on our side as you can see in the second graph I posted:
![39815219-96da718e-5397-11e8-9112-97631d8e1e87](https://user-images.githubusercontent.com/14098361/40825576-40d60274-6578-11e8-9551-9730d029f6cf.png)
This shows a metric (database CPU usage) extracted directly from the database (not from datadog), and the shape is roughly the same for the first, second and third wave.

But on the first graph I posted, you see that when we updated the gem there was a huge increase in the datadog metric supposed to measure number of requests to the database (see second wave, and also third wave which is clearly higher than the first one whereas it is the not the case on the above graph):
![39815221-9855169a-5397-11e8-8776-55d55754260d](https://user-images.githubusercontent.com/14098361/40825748-d9e413b6-6578-11e8-886f-a68377421cd6.png)

So basically the patterns of the 2 metrics are in sync, fortunately, but this increase made us worry we deployed a very wrong version of our code or had a huge traffic increase that we didn't foresee. One day we were at 600 req/s max and the next, it went up to 1200 req/s: obviously, only one of these 2 values can reflect what is really happening.
When you say that from 0.7.x to 0.11.x you are tracing a lot more things, does this means that previously you were not tracing all queries to the DB from ActiveRecord and that now you do? Or do you still trace only a subset of AR DB queries and I shouldn't take this rps indicator too seriously?

**Comment by delner** [comment]
@Mathiou04 Okay, glad you're not experiencing any kind of performance issues. Off the top of my head, I can't really explain all the differences in traces between 0.7.x and 0.11.x (I've only been working on the project since 0.10.x), but that it's the most likely explanation for the big jump.

What I can offer though is that the agent is responsible for calculating this RPS metric, and it does so by counting the number of traces it sees per service. So the number should be correlated with an increase in the number of traces you're receiving for a particular service (but maybe not with "requests", exactly.)

**Comment by delner** [comment]
Apologies if this isn't as thorough an explanation that you were hoping for, but is there anything else I can offer for this issue?

**Comment by Mathiou04** [comment]
@delner Yes, I understand there was an increase of traces received for this particular service, I just wanted to ensure that the new version of the gem give metrics closer to reality than the old one and is not sending duplicated traces or something like that.
I hope it was just a bugfix to ensure all DB calls are traced and that it will stay stable in the future :)
Thanks for having a look anyway.

**Comment by delner** [comment]
@Mathiou04 Unfortunately I can't really provide a more thorough explanation for what you saw in your application, but I can say we have no reason to believe the RPS numbers aren't accurate now.

Is it okay if I close this one?

**Comment by Mathiou04** [comment]
Yes, it is fine with me.
